### PR TITLE
Add gnorm metric logging when gradient clipping < 0

### PR DESCRIPTION
**Pull request by hadasah**
**Patch description**
Patch for #1977
Show gnorm in metrics even when gradient clipping isn't set

**Testing steps**
Ran train_model with --gradient_clip -1, verified that gnorm is printed.
Command:
`python examples/train_model.py -m transformer/generator -t convai2 -mf /tmp/tg --gradient_clip -1 -lr 1e-10`
Output:
Screen Shot 2020-03-25 at 17 43 04

### CODE REVIEW

File: parlai/utils/torch.py [hunk 1]
```diff
@@ -226,6 +226,29 @@ def argsort(keys: List[Any], *lists: List[List[Any]], descending: bool = False):
...
+        the model parameters for gradient norm calculation. Iterable of
+        Tensors or single Tensor
+
+    :param norm_type
+        type of p-norm to use
+
+    """
+    if isinstance(parameters, torch.Tensor):
+        parameters = [parameters]
+    parameters = list(filter(lambda p: p.grad is not None, parameters))
```
Comment (stephenroller) [inline parlai/utils/torch.py]:
```suggestion
    parameters = [p.grad for p in parameters if p is not None]
```

File: parlai/utils/torch.py [hunk 2]
```diff
@@ -226,6 +226,29 @@ def argsort(keys: List[Any], *lists: List[List[Any]], descending: bool = False):
...
+
+    :param parameters:
+        the model parameters for gradient norm calculation. Iterable of
+        Tensors or single Tensor
+
+    :param norm_type
+        type of p-norm to use
+
+    """
+    if isinstance(parameters, torch.Tensor):
```
Comment (stephenroller) [inline parlai/utils/torch.py]: I don't understand this one

Comment (hadasah) [reply parlai/utils/torch.py]: torch.nn.utils.clip_grad_norm_ works in the case where it's passed a list of tensors, as well as a single tensor of parameters. Since I don't think we'll be needing that case (?) I can just get rid of that

File: parlai/utils/torch.py (hunk 1 above)
Comment (hadasah) [reply parlai/utils/torch.py]: oops I forgot to change this, thanks!

File: parlai/utils/torch.py (hunk 2 above)
Comment (stephenroller) [reply parlai/utils/torch.py]: Oh, that's cool. I like the mirroring, so you can leave it if you want.

Do you know what fixed the tests?

File: parlai/utils/torch.py [hunk 3]
```diff
@@ -226,6 +226,29 @@ def argsort(keys: List[Any], *lists: List[List[Any]], descending: bool = False):
     return output
 
 
+def compute_grad_norm(parameters, norm_type=2.0):
+    """
+    Compute norm over gradients of model parameters.
```
Comment (stephenroller) [inline parlai/utils/torch.py]: It wants an extra newline between this and the next line, and it wants you to delete the extra empty line at the end.

File: parlai/utils/torch.py [hunk 4]
```diff
@@ -476,9 +501,13 @@ def chunk_to(chunk: Chunk, device: str) -> Chunk:
         if isinstance(chunk, torch.Tensor):
             return chunk.to(device)  # type: ignore
         elif isinstance(chunk, tuple):
-            return tuple(PipelineHelper.chunk_to(c, device) for c in chunk)  # type: ignore
+            return tuple(
+                PipelineHelper.chunk_to(c, device) for c in chunk
+            )  # type: ignore
         elif isinstance(chunk, dict):
-            return {k: PipelineHelper.chunk_to(v, device) for k, v in chunk.items()}  # type: ignore
```
Comment (stephenroller) [inline parlai/utils/torch.py]: please don't change any of these?

**Comment by hadasah** [comment]
> I wonder if there’s a test we could write....

I could write a function in test_utils_torch.py just for testing the compute_grad_norm() function but I don't think we currently test for which metrics get printed out during training (do we? I didn't find it but maybe I wasn't looking in the right place)

**Comment by stephenroller** [comment]
Yeah that's correct. Go ahead and merge.

What was wrong with the docs?

**Comment by hadasah** [comment]
it didn't like the reference to torch.nn.utils.clip_grad_norm_ - I think it tries to link to the functioin so you can only reference functions within ParlAI?
